### ISSUE TITLE
Deadlock in GIOPConnection.close()

### ISSUE THREAD

**Issue by jagello** [JacORB/JacORB#206]
I am using JacORB v3.8 with bidirectional connections.
Recently I experienced a deadlock on server side. Looks like 2 threads were trying to close the same ServerGIOPConnection and since GIOPConnection.close() needs to lock 2 objects (one is the Object connect_sync, second is the ClientConnectionManager on calling ClientConnectionManager.removeConnection()).
The 2 threads were obtaining the locks for these 2 objects in different order, so the deadlock was alive.

To me it looks like thread DBChangeNotificator is closing the connection after a successful request, but RequesProcessor-49 is closing it due to on error in communication.

Pasting the stacktraces of the 2 threads locking each other.

Name: RequestProcessor-49 
State: BLOCKED on org.jacorb.orb.giop.ClientConnectionManager@2f2ec593 owned by: DBChangeNotificator 
Total blocked: 32,122 Total waited: 8,121
 
Stack trace: 
org.jacorb.orb.giop.ClientConnectionManager.removeConnection(ClientConnectionManagerjava:192) 
org.jacorb.orb.giop.ClientConnection.connectionClosed(ClientConnection.java:455)
org.jacorb.orb.giop.GIOPConnection.close(GIOPConnection.java:1170) 

`- locked java.lang,Object@277a795d `

org.jacorb.orb.giop.ServerGIOPConnection.close(ServerGIOPConnection.java:237) 
org.jacorb.orb.giop.ServerGIOPConnection.streamClosed(ServerGIOPConnection.java:229) 
org.jacorb.orb.giop.GIOPConnection.sendMessage(GIOPConnection.java:1143) 
org.jacorb.orb.giop.GIOPConnection.sendMessage(GIOPConnection.java:1034)
org.jacorb.orb.giop.GIOPConnection.sendReply(GIOPConnection.java:1028)
org.jacorb.orb.dsi.ServerRequest.reply(ServerRequest.java:456)
org.jacorb.orb.BasicAdapter.return_result(BasicAdapterjava:375)
org.jacorb.poa.RequestController.returnResult(RequestController.java:398) 
org.jacorb.poa.RequestProcessor.run(RequestProcessorjava:885) 

Name: DBChangeNotificator
State: BLOCKED on java.lang.Object@277a795d owned by: RequestProcessor-49 
Total blocked: 139 Total waited: 124 

Stack trace: 
org.jacorb.orb.giop.GIOPConnection.close(GIOPConnection.java:1168) 
org.jacorb.orb.giop.ServerGIOPConnection.close(ServerGIOPConnection.java:237) 
org.jacorb.orb.giop.ClientConnection.close(ClientConnection.java:329) 
org.jacorb.orb.giop.ClientConnectionManager.releaseConnection(ClientConnectionManagerjava:171) 
`- locked org.jacorbeorb.giop.ClientConnectionManager@2f2ec593 `
org.jacorb.orb.Delegate._invoke_internal(Delegate,java:1496) 
org.jacorb.orb.Delegate.invoke_internal(Delegate.java:1209) 
org.jacorb.orb.Delegate.invoke(Delegate.java:1197) 
org.omg.CORBA.portable.ObjectImpLinvoke(ObjectImpliava:475) 
eu.extech.quant.client._QuantClientStub.recieveDBChangeNotificationList(Unknown Source)

**Comment by rnc** [comment]
Thanks for the report. Have you managed to reproduce this in a test (otherwise I guess we'll have to use e.g. Byteman e,g, https://github.com/JacORB/JacORB/blob/master/test/regression/src/test/java/org/jacorb/test/bugs/bug1002/Bug1002Test.java for a test case),

**Comment by jagello** [comment]
No, I did not reproduce it in a test case. I believe that a race-condition is needed between the 2 threads executing 2 separate requests, so I can not really think of a test case which will generate the dead lock on each run. If I needed to reproduce it, I would try debug mode with some breakpoints at the ServerGIOPConnection.close() method.

**Comment by rnc** [comment]
Are you using BiDirectionional connections ? Do you have any debug logs from the case you saw? 

It might be feasible to adjust the locking within ClientConnectionManager::releaseConnection so that it doesn't lock on the method but on the connections collection. This should avoid this scenario.

**Comment by jagello** [comment]
Yes, I am using bidirectional connections. I don't have a debug log, because I saw this incident only once, but I will try to reproduce it with debug log using some breakpoints.

**Comment by chylek-extech** [comment]
I was able to reproduce the issue by introducing a delay in `ClientConnectionManager.releaseConnection` (even a 10ms delay was about 80-90% chance of deadlock), and triggering a `COMM_FAILURE` in `GIOPConnection.sendMessage`:

```diff
Index: core/src/main/java/org/jacorb/orb/giop/ClientConnectionManager.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- core/src/main/java/org/jacorb/orb/giop/ClientConnectionManager.java	(revision f0549593d0a2f44985caa64e1fc44aa343a9f7d2)
+++ core/src/main/java/org/jacorb/orb/giop/ClientConnectionManager.java	(date 1603269216245)
@@ -162,6 +162,11 @@
     {
         if ( connection.decClients() || connection.isClosed ())
         {
+            try {
+                Thread.sleep(50L);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
             if (logger.isDebugEnabled())
             {
                 logger.debug ("ClientConnectionManager: releasing " +
Index: core/src/main/java/org/jacorb/orb/giop/GIOPConnection.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- core/src/main/java/org/jacorb/orb/giop/GIOPConnection.java	(revision f0549593d0a2f44985caa64e1fc44aa343a9f7d2)
+++ core/src/main/java/org/jacorb/orb/giop/GIOPConnection.java	(date 1603269175397)
@@ -27,6 +27,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.locks.ReentrantLock;
 import org.jacorb.config.Configuration;
 import org.jacorb.config.ConfigurationException;
@@ -1033,9 +1034,11 @@
         sendMessage (out, null);
     }
 
+    public static final AtomicBoolean COMM_ERR_MANUAL = new AtomicBoolean(false);
     private final void sendMessage( MessageOutputStream out, org.omg.TimeBase.UtcT sendDeadline)
         throws IOException
     {
+        boolean commErr = COMM_ERR_MANUAL.get();
         try
     	{
             try
@@ -1091,6 +1094,7 @@
 
                 out.write_to( this );
                 transport.flush();
+                if (commErr) throw new COMM_FAILURE("intentional after transport flush");
 
                 if (logger.isDebugEnabled())
                 {
```
It should be possible to reproduce this reliably by using a bidirectional connection, and enabling `COMM_ERR_MANUAL` for a short while (say, 500 ms) right when a client request comes in, before a reply is sent back. Note that it is crucial that there is only 1 client, so that `decClients` succeeds and goes into the first branch where the delay is.

The patch above was created against the most recent commit in master (`f0549593d0a2f44985caa64e1fc44aa343a9f7d2`), but please note that we are currently stuck on v3.8 because of other issues. I noticed the only change to these 2 classes between v3.8 and master was exception handling, which shouldn't by itself affect this issue.

Your suggestion of locking on the `connections` collection (on all synchronized methods) unfortunately did not fix the deadlock alone. What does appear to work is synchronizing on `connections` *and* calling `connection.close()` outside the synchronized block. I do not believe moving `close()` will have any adverse effects, but please correct me if I'm wrong.
```java
    public void releaseConnection( ClientConnection connection )
    {
        boolean closing = false;
        synchronized (connections) {
            if (connection.decClients() || connection.isClosed()) {
                if (logger.isDebugEnabled()) {
                    logger.debug("ClientConnectionManager: releasing " +
                            (connection.isClosed() ? "closed connection " : "") +
                            connection.getGIOPConnection().toString());
                }
                closing = true;
                connections.remove(connection.getRegisteredProfile());
            } else {
                // not sure if this should be a warning or even an error
                if (logger.isDebugEnabled()) {
                    logger.debug("ClientConnectionManager: cannot release "
                            + connection.getGIOPConnection().toString()
                            + " (still has " + connection.numClients() + " client(s))");
                }
            }
        }
        if (closing) {
            connection.close();
        }
    }
```

**Comment by goalan-ms** [comment]
I believe we've also hit this deadlock.
If the suggested fix is good, is there any chance it could be merged?
Is there anything I could do to help with that?
